datatypes/src/primitives/datetime.rs

```diff
@@ -0,0 +1,668 @@
+    /// Assumes the time is in UTC.
+    ///
+    /// If the date or the time would overflow, it returns `None`
+    ///
+    pub fn new_utc_checked_with_millis(
+        year: i32,
+        month: u8,
+        day: u8,
+        hour: u8,
+        minute: u8,
+        second: u8,
+        millis: u16,
+    ) -> Option<Self> {
+        let date = NaiveDate::from_ymd_opt(year, month.into(), day.into())?;
+        let datetime =
+            date.and_hms_milli_opt(hour.into(), minute.into(), second.into(), millis.into())?;
+
+        Some(Self {
+            datetime,
+            offset: UTC,
+        })
+    }
+
+    #[must_use]
+    pub fn into_utc(self) -> Self {
+        if self.offset == UTC {
+            return self;
+        }
+
+        Self {
+            datetime: self.datetime + self.offset._to_fixed_offset(),
+            offset: UTC,
+        }
+    }
+
+    pub fn parse_from_str(
+        input: &str,
+        format: &DateTimeParseFormat,
+    ) -> Result<Self, DateTimeError> {
+        match (format.has_time(), format.has_tz()) {
+            (true, true) => {
+                let datetime = chrono::DateTime::<chrono::FixedOffset>::parse_from_str(
+                    input,
+                    format._to_parse_format(),
+                )
+                .map_err(|e| DateTimeError::DateParse {
+                    source: Box::new(e),
+                })?;
+
+                Ok(Self {
+                    datetime: datetime.naive_utc(),
+                    offset: UtcOffset::_from_fixed_offset(datetime.offset().fix()),
+                })
+            }
+            (true, false) => {
+                let datetime =
+                    chrono::NaiveDateTime::parse_from_str(input, format._to_parse_format())
+                        .map_err(|e| DateTimeError::DateParse {
+                            source: Box::new(e),
+                        })?;
+
+                Ok(Self {
+                    datetime,
+                    offset: UTC,
+                })
+            }
+            (false, true) => Err(DateTimeError::CannotParseOnlyDateWithTimeZone),
+            (false, false) => {
+                let datetime = chrono::NaiveDate::parse_from_str(input, format._to_parse_format())
+                    .map_err(|e| DateTimeError::DateParse {
+                        source: Box::new(e),
+                    })?
+                    .and_hms(0, 0, 0);
+
+                Ok(Self {
+                    datetime,
+                    offset: UTC,
+                })
+            }
+        }
+    }
+
+    pub fn parse_from_rfc3339(input: &str) -> Result<Self, DateTimeError> {
+        let date_time = chrono::DateTime::<chrono::FixedOffset>::parse_from_rfc3339(input)
+            .map_err(|e| DateTimeError::DateParse {
+                source: Box::new(e),
+            })?;
+
+        Ok(date_time.into())
+    }
+
+    // TODO: is this inprecise?
```

Comment:
what does the precision refer to? format or temporal resolution?

Reply:
We just found that in chrono `from_str` parses stuff with slightly more slack that rfcXYZ.
For instance, we allow negative date times.
But maybe we need just a definition of what we exactly want and then implement that?